resources: Refactor arm and x86 disk image directories

Harshil2107 opened the pull request:
- This change merges the packer files and adds an command line argument which can be used to specify a version of ubuntu to build.

- Update build.sh to use the merged packer file

- Merge the arm and x86 disk image directories into a single directory

Harshil2107 committed:
resources: Merge packer files for x86 img and arm images

- This change moves the files to build arm and x86 images to a common directory

- This change merges the packer files and adds an command line argument which can be used to specify a version of ubuntu to build.

- Update build.sh to use the merged packer file

- updates documentation for building and running the diskiamges

powerjg commented on src/ubuntu-generic-diskimages/scripts/post-installation.sh:
```diff
@@ -45,9 +51,9 @@ cp -r include/gem5 /usr/local/include/\
+scons build/${ISA}/out/m5
+cp build/${ISA}/out/m5 /usr/local/bin/
+cp build/${ISA}/out/libm5.a /usr/local/lib/
 popd
 popd
```
It would be good to let people know what we're popping

```suggestion
popd   # util/m5
popd   # gem5
```

powerjg commented on src/ubuntu-generic-diskimages/build.sh:
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
```
This file should be named `build-x86.sh`

powerjg commented on src/ubuntu-generic-diskimages/build-arm.sh:
```diff
@@ -7,7 +7,7 @@ if [ ! -f ./packer ]; then
 fi
```
The code above this is a problem. It's specific for arm *hosts*. 

So, we need to make sure that we differentiate between Arm hosts and Arm guest.

powerjg reviewed (changes requested):
A few high-level things that aren't included in the review below.
1. `BUILDING.md` and `README.md` are x86-specific. These need to be updated.
2. What is `flash0.img`? Do we have to have it in git? It's Arm-specific (I assume), so maybe it should be in `files/arm`
3. There could be confusion between Arm host and Arm guest. This needs to be both documented well and we need to make sure that we aren't doing anything wrong.

powerjg commented on src/ubuntu-generic-diskimages/BUILDING.md:
````diff
@@ -88,6 +92,17 @@ The `user-data` file in this repo, is made by selecting all default options exce
+
+We need a EFI file to boot the arm image. We use the file named `flash0.img` in the packer file.
+
+To get the `flash0.img` run the following commands in the `files` directory
+
+```bash
+dd if=/dev/zero of=flash0.img bs=1M count=64
+dd if=/usr/share/qemu-efi-aarch64/QEMU_EFI.fd of=flash0.img conv=notrunc
````
Why can't you just use `/usr/share/qemu-efi-aarch64/QEMU_EFI.fd`? Why do you have to copy that file in such a strange way?

Harshil2107 replied:
We do this because we need it to be 64MB with 0s if the file is smaller.

powerjg commented on src/ubuntu-generic-diskimages/BUILDING.md:
```diff
@@ -4,18 +4,22 @@ authors:
 - `scripts`: Scripts run on the disk image after installation.
-- `http`: cloud-init Ubuntu autoinstall files.
+- `http`: cloud-init Ubuntu autoinstall files for ubuntu 22.04.
+- `http-24.04`: cloud-init Ubuntu autoinstall files for ubuntu 24.04.
```
What about x86 24.04? There's no files for that.

Harshil2107 replied:
We dont have separate user-data for x86 22.04 and 24.04

powerjg commented on src/ubuntu-generic-diskimages/build-arm.sh:
Needs copyright, but it could be the simple SPDX block, not the full thing.

powerjg commented on src/ubuntu-generic-diskimages/README.md:
This definitely needs to be updated. It says "x86" and "22.04" in many places.

powerjg commented on src/ubuntu-generic-diskimages/x86-ubuntu-24-04.pkr.hcl:
Why is this not in `packer-scripts`?

powerjg commented on src/ubuntu-generic-diskimages/packer-scripts/x86-ubuntu.pkr.hcl:
```diff
@@ -22,6 +22,30 @@ variable "ssh_username" {
 
+variable "ubuntu_version" {
+  type    = string
+  default = "22.04"
```
Change the default to 24.04

powerjg commented on src/ubuntu-generic-diskimages/packer-scripts/x86-ubuntu.pkr.hcl:
```diff
@@ -22,6 +22,30 @@ variable "ssh_username" {
+    "22.04" = {
+      iso_url       = "https://old-releases.ubuntu.com/releases/jammy/ubuntu-22.04.2-live-server-amd64.iso"
+      iso_checksum  = "sha256:5e38b55d57d94ff029719342357325ed3bda38fa80054f9330dc789cd2d43931"
+      output_dir    = "disk-image"
```
Change this to `disk-image-22.04`

powerjg commented on src/ubuntu-generic-diskimages/packer-scripts/arm-ubuntu.pkr.hcl:
```diff
@@ -22,6 +22,32 @@ variable "ssh_username" {
+      iso_url       = "https://old-releases.ubuntu.com/releases/jammy/ubuntu-22.04-live-server-arm64.iso"
+      iso_checksum  = "sha256:c209ab013280d3cd26a344def60b7b19fbb427de904ea285057d94ca6ac82dd5"
+      output_dir    = "disk-image"
+      http_directory = "http/arm"
```
Change this to `http-22-04/arm

powerjg reviewed (commented):
I've made some comments below, but my comments are not complete. You need to improve the consistency throughout. Don't treat arm and x86 differently. Don't treat 22.04 and 24.04 differently. E.g., if for one you call it `x86-22.04` then you should also use `arm-24.04`. *Be consistent*

powerjg commented on src/ubuntu-generic-diskimages/BUILDING.md:
```diff
@@ -43,29 +53,28 @@ The kernel is extracted using packer's file provisioner with `direction=download
   If you require the service to start immediately without waiting for the next boot then also run the following:
   `sudo systemctl start systemd-networkd-wait-online.service`.
 
 ## Extending the Disk Image
```
I think you should remove this line. You are not describing *how* to extend the disk image. You are describing the changes we have made to the disk image.

powerjg commented on src/ubuntu-generic-diskimages/BUILDING.md:
```diff
@@ -43,29 +53,28 @@ The kernel is extracted using packer's file provisioner with `direction=download
 
-- **Replace `gem5_init.sh`**: This script is what executes as the Linux init process (pid=0) immediately after Linux boot. If you have a custom initialization script, replace the default `gem5_init.sh` in both `x86-ubuntu.pkr.hcl` and `post-installation.sh` to integrate your custom initialization process.
-- **Replace `gem5_init.sh`**: If you have a custom initialization script, replace the default `gem5_init.sh` in both `x86-ubuntu.pkr.hcl` and `post-installation.sh` to integrate your custom initialization process.
+- **Replace `gem5_init.sh`**: This script is what executes as the Linux init process (pid=0) immediately after Linux boot. If you have a custom initialization script, replace the default `gem5_init.sh` in the packer file that you are using such as the `x86-ubuntu.pkr.hcl` and `post-installation.sh` to integrate your custom initialization process.
```
```suggestion
- **`gem5_init.sh` replaces /sbin/init**: This script is what executes as the Linux init process (pid=0) immediately after Linux boot. This script....
```

powerjg commented on src/ubuntu-generic-diskimages/BUILDING.md:
````diff
@@ -75,6 +84,21 @@ To avoid its infinite execution, we incorporated a conditional check in `post-in
   chmod u+s /path/to/gem5-bridge
   ```
 
+### Adding more libraries
````
```suggestion
## Extending the disk image with custom files and scripts
```

powerjg commented on src/ubuntu-generic-diskimages/BUILDING.md:
```diff
@@ -75,6 +84,21 @@ To avoid its infinite execution, we incorporated a conditional check in `post-in
 
+### Adding more libraries
+
+- You can add more libraries to the disk image by updating the `post-installation.sh` script.
```
```suggestion
- You can add more packages to the disk image by updating the `post-installation.sh` script.
```

powerjg commented on src/ubuntu-generic-diskimages/BUILDING.md:
```diff
@@ -75,6 +84,21 @@ To avoid its infinite execution, we incorporated a conditional check in `post-in
+
+If you need to increase the size of the image when adding more libraries and files to the image update the size of partition in the respective `user-data` file. Also, update the `disk_size` parameter to be atleast one mega byte more than
+the size you defined in the `user-data` file.
+
```
```suggestion
> **NOTE:** You can extend this disk image by modifying the `post-installation.sh` script, but it requires building the image from scratch.
> To take a pre-built image and add new files or packages, see examples in ....
```

powerjg commented on src/ubuntu-generic-diskimages/README.md:
```diff
@@ -21,7 +20,7 @@ Thus, you should be able to build packages on the disk and easily link to the ge
 The disk has network disabled by default to improve boot time in gem5.
 
-If you want to enable networking, you need to modify the disk image and move the file `/etc/netplan/00-installer-config.yaml.bak` to `/etc/netplan/00-installer-config.yaml`.
+If you want to enable networking, you need to modify the disk image and move the file `/etc/netplan/00-installer-config.yaml.bak` or `/etc/netplan/50-cloud-init.yaml.bak` to `/etc/netplan/00-installer-config.yaml` or `/etc/netplan/50-cloud-init.yaml` depending on which config file the disk image contains.
```
Add something like "See ... for an example of how to do this"

powerjg commented on src/ubuntu-generic-diskimages/README.md:
````diff
@@ -116,16 +115,38 @@ scons build/X86/gem5.opt -j`nproc`
 ```sh
 scons build/X86/gem5.opt -j`nproc`
-./build/X86/gem5.opt configs/example/gem5_library/x86-ubuntu-run-with-kvm.py`
+./build/X86/gem5.opt configs/example/gem5_library/x86-ubuntu-run-with-kvm.py
````
```suggestion
./build/ALL/gem5.opt configs/example/gem5_library/x86-ubuntu-run-with-kvm.py
```

Let's stop using the ISA-specific builds.

Harshil2107 committed:
resources: Update documentation

erin-le commented on src/ubuntu-generic-diskimages/BUILDING.md:
```diff
@@ -1,21 +1,30 @@
 - `scripts`: Scripts run on the disk image after installation.
-- `http`: cloud-init Ubuntu autoinstall files.
-- `disk-image`: The disk image output directory.
+- `http`: cloud-init Ubuntu autoinstall files for different versions of ubuntu for arm and x86.
```
```suggestion
- `http`: cloud-init Ubuntu autoinstall files for different versions of Ubuntu for Arm and x86.
```

erin-le commented on src/ubuntu-generic-diskimages/BUILDING.md:
```diff
@@ -43,29 +52,26 @@ The kernel is extracted using packer's file provisioner with `direction=download
   - If the `interactive` boot option is passed, the `gem5-bridge exit` command is not run after the linux kernel initialization.
-- Networking is disabled by moving the `/etc/netplan/00-installer-config.yaml` file to `/etc/netplan/00-installer-config.yaml.bak`. The `systemd-networkd-wait-online.service` is also disabled.
-  - If you want to enable networking, you need to modify the disk image and move the file `/etc/netplan/00-installer-config.yaml.bak` to `/etc/netplan/00-installer-config.yaml`.
+- Networking is disabled by moving the `/etc/netplan/00-installer-config.yaml` or `/etc/netplan/50-cloud-init.yaml` file to `/etc/netplan/00-installer-config.yaml.bak` or `/etc/netplan/50-cloud-init.yaml.bak` respectively. The `systemd-networkd-wait-online.service` is also disabled.
```
Clarify which .yaml corresponds with which category of disk images.

erin-le commented on src/ubuntu-generic-diskimages/README.md:
```diff
@@ -21,7 +20,12 @@ Thus, you should be able to build packages on the disk and easily link to the ge
 The disk has network disabled by default to improve boot time in gem5.
 
-If you want to enable networking, you need to modify the disk image and move the file `/etc/netplan/00-installer-config.yaml.bak` to `/etc/netplan/00-installer-config.yaml`.
+If you want to enable networking, you need to modify the disk image and move the file `/etc/netplan/00-installer-config.yaml.bak` or `/etc/netplan/50-cloud-init.yaml.bak` to `/etc/netplan/00-installer-config.yaml` or `/etc/netplan/50-cloud-init.yaml` depending on which config file the disk image contains. For example you can use the following commands to re-enable network:
```
Clarify which disk images these belong to

erin-le commented on src/ubuntu-generic-diskimages/BUILDING.md:
```diff
@@ -25,13 +34,13 @@ See [Troubleshooting](#troubleshooting) for more information.
-The latest headers and modules are installed using apt, before extracting the kernel using the `extract-vmlinux` script provided in ubuntu. The extracted kernel is placed at `/home/gem5/vmlinux-x86-ubuntu`.
+For the x86 disk images a kernel is also extracted from the disk image during the post-installation process.
+The latest headers and modules are installed using apt, before extracting the kernel using the `extract-vmlinux` script provided in ubuntu. The extracted kernel is placed at `/home/gem5/vmlinux-x86-ubuntu` in the disk image.
 The extracted kernel does not have a version its name, but the kernel version is printed as before the extraction in `post-installation.sh` script. This extracted kernel can be used as a resource for gem5 simulations and is not limited to just be used with this disk image.
```
```suggestion
The extracted kernel does not have a version in its name, but the kernel version is printed before the extraction in `post-installation.sh` script. This extracted kernel can be used as a resource for gem5 simulations and is not limited to just be used with this disk image.
```

erin-le commented on src/ubuntu-generic-diskimages/BUILDING.md:
```diff
@@ -43,29 +52,26 @@ The kernel is extracted using packer's file provisioner with `direction=download
   - If the `no_systemd` boot option is passed, systemd is not run and the user is dropped to a terminal.
   - If the `interactive` boot option is passed, the `gem5-bridge exit` command is not run after the linux kernel initialization.
-- Networking is disabled by moving the `/etc/netplan/00-installer-config.yaml` file to `/etc/netplan/00-installer-config.yaml.bak`. The `systemd-networkd-wait-online.service` is also disabled.
-  - If you want to enable networking, you need to modify the disk image and move the file `/etc/netplan/00-installer-config.yaml.bak` to `/etc/netplan/00-installer-config.yaml`.
```
Clarify which disk images these files correspond to

erin-le commented on src/ubuntu-generic-diskimages/BUILDING.md:
````diff
@@ -75,6 +81,25 @@ To avoid its infinite execution, we incorporated a conditional check in `post-in
+  }
+```
+
+If you need to increase the size of the image when adding more libraries and files to the image update the size of partition in the respective `user-data` file. Also, update the `disk_size` parameter to be atleast one mega byte more than
````
```suggestion
If you need to increase the size of the image when adding more libraries and files to the image update the size of the partition in the respective `http/*/user-data` file. Also, update the `disk_size` parameter in `post-installation.sh` to be at least one mega byte more than
```

erin-le commented on src/ubuntu-generic-diskimages/BUILDING.md:
```diff
@@ -75,6 +81,25 @@ To avoid its infinite execution, we incorporated a conditional check in `post-in
+
+**NOTE:** You can extend this disk image by modifying the `post-installation.sh` script, but it requires building the image from scratch.
+
+To take a pre-built image and add new files or packages, see the post in gem5 discussion.
```
Add link to the gem5 discussion

Harshil2107 replied:
I added the link to the documentation but it is still in PR form, the link should work as soon as it is merged.

erin-le commented on src/ubuntu-generic-diskimages/README.md:
````diff
@@ -104,28 +108,50 @@ By selecting the appropriate parameters, users can customize the boot process fo
+./build/ALL/gem5.opt configs/example/gem5_library/x86-ubuntu-run-with-kvm.py
 ```
 
 **Note:** the `x86-ubuntu-with-kvm.py` script requires a host machine with KVM to function correctly.
````
```suggestion
**Note:** the `x86-ubuntu-with-kvm.py` script requires a x86 host machine with KVM to function correctly.
```

erin-le commented on src/ubuntu-generic-diskimages/README.md:
````diff
@@ -104,28 +108,50 @@ By selecting the appropriate parameters, users can customize the boot process fo
+./build/ALL/gem5.opt configs/example/gem5_library/arm-ubuntu-run-with-kvm.py
+```
+
+**Note:** the `arm-ubuntu-run-with-kvm.py` script requires a host machine with KVM to function correctly.
````
```suggestion
**Note:** the `arm-ubuntu-run-with-kvm.py` script requires an Arm host machine with KVM to function correctly.
```

Harshil2107 committed:
resources: update packer script of arm

- use concat function to add qemu args
- remove arm-ubuntu directory

erin-le commented on src/ubuntu-generic-diskimages/BUILDING.md:
````diff
@@ -88,6 +113,17 @@ The `user-data` file in this repo, is made by selecting all default options exce
+To get the `flash0.img` run the following commands in the `files` directory
+
+```bash
+dd if=/dev/zero of=flash0.img bs=1M count=64
````
Can the commands to get flash0.img be incorporated into build-arm.sh and/or be mentioned in the "## Disk Image" section of the document? The commands are listed after information about how to customize the disk image, so it may be difficult for users to figure out that they need to run them.

Harshil2107 replied:
Moved the section to the `## Disk Image` section.

erin-le commented on src/ubuntu-generic-diskimages/BUILDING.md:
```diff
@@ -1,21 +1,30 @@
-Run `./build.sh` in the `x86-ubuntu` directory to build the disk image.
+Run `build-x86.sh` with the argument `22.04` or `24.04` to build the respective x86 disk image in the `ubuntu-generic-diskimages` directory.
+Run `build-arm.sh` with the argument `22.04` or `24.04` to build the respective arm disk image in the `ubuntu-generic-diskimages` directory.
+Building the arm image assume that we are on an ARM machine as we use kvm to build the image.
```
Add info about how to build Arm disk images without using KVM

erin-le commented on src/ubuntu-generic-diskimages/README.md:
````diff
@@ -104,28 +108,50 @@ By selecting the appropriate parameters, users can customize the boot process fo
+```sh
+scons build/ARM/gem5.opt -j `nproc`
+./build/ALL/gem5.opt configs/example/gem5_library/arm-ubuntu-run-with-kvm.py
+```
````
These example scripts show how to run the workload and not how to set up a simulation with a disk image, kernel, and bootloader. I think it is helpful to point the user to the example scripts, but you should also add instructions on how to use `DiskImageResource` and `set_kernel_disk_workload` with the disk image the user built.

erin-le replied:
Here's the chunk of code I used for Arm:

```
disk_img = DiskImageResource("/path/to/disk/image/directory/arm-disk-image-22-04/arm-ubuntu", root_partition="2")
board.set_kernel_disk_workload(
        disk_image=disk_img,
        bootloader=obtain_resource("arm64-bootloader-foundation"),
        kernel=obtain_resource("arm64-linux-kernel-5.15.36")
    )
```

Harshil2107 committed:
resources: Address requested changes

Harshil2107 committed:
resources: Update after boot and gem5_init to use address version of m5ops

Harshil2107 committed:
resources: remove -vga qemu arg

- removing this argument as it was causing the packer script to fail and removing this argument does not change anything about the disk image.

Harshil2107 committed:
resources: update documentation

erin-le commented on src/ubuntu-generic-diskimages/build-arm.sh:
```diff
@@ -0,0 +1,36 @@
+    echo "Error: Invalid Ubuntu version '$ubuntu_version'. Must be '22.04' or '24.04'."
+    exit 1
+fi
+
```
```suggestion

# Get flash0.img
cd files

dd if=/dev/zero of=flash0.img bs=1M count=64
dd if=/usr/share/qemu-efi-aarch64/QEMU_EFI.fd of=flash0.img conv=notrunc

cd ..

```

Harshil2107 committed:
resources: update build-arm.sh to make the flash0.img file

erin-le reviewed (dismissed):
LGTM. I tested building and using the disk images as resources in a simulation and they all work. Thanks for all of the changes!

powerjg commented on src/ubuntu-generic-diskimages/scripts/post-installation.sh:
```diff
@@ -52,11 +58,11 @@ cp -r include/gem5 /usr/local/include/\
+popd   # gem5
 
 # rename the m5 binary to gem5-bridge
 mv /usr/local/bin/m5 /usr/local/bin/gem5-bridge
```
Line 80 below is wrong. It says "x86-ubuntu.pkr.hcl"

powerjg commented on src/ubuntu-generic-diskimages/README.md:
```diff
@@ -21,7 +20,13 @@ Thus, you should be able to build packages on the disk and easily link to the ge
 The disk has network disabled by default to improve boot time in gem5.
 
-If you want to enable networking, you need to modify the disk image and move the file `/etc/netplan/00-installer-config.yaml.bak` to `/etc/netplan/00-installer-config.yaml`.
+If you want to enable networking, you need to modify the disk image and move the file `/etc/netplan/00-installer-config.yaml.bak` or `/etc/netplan/50-cloud-init.yaml.bak` to `/etc/netplan/00-installer-config.yaml` or `/etc/netplan/50-cloud-init.yaml` depending on which config file the disk image contains. The x86 ubuntu 22.04 image should have `/etc/netplan/00-installer-config.yaml` and the other ...
```
Above this on line 17, it says "The headers for `libm5` are installed in `/usr/local/include/gem5-bridge`" that should be "The headers for `libm5` are installed in `/usr/local/include`"

Harshil2107 committed:
resoucres: Address requested changes
